concordium-consensus/src/Concordium/KonsensusV1/TreeState/LowLevel/LMDB.hs

```diff
@@ -504,55 +648,250 @@ instance
+                        _ <-
+                            deleteRecord
+                                txn
+                                (dbh ^. latestFinalizationEntryStore)
+                                CSKLatestFinalizationEntry
+                        return (c, hashes, fin)
+                    NormalBlock block -> do
+                        -- Remove the block and its transactions
+                        let finHash = getHash fin
+                        _ <- deleteRecord txn (dbh ^. blockStore) finHash
+                        _ <- deleteRecord txn (dbh ^. finalizedBlockIndex) (blockHeight fin)
+                        forM_ (blockTransactions fin) $
+                            deleteRecord txn (dbh ^. transactionStatusStore) . getHash
+                        mparent <- loadRecord txn (dbh ^. blockStore) (blockParent block)
+                        case mparent of
+                            Nothing ->
+                                throwM . DatabaseInvariantViolation $
+                                    "Missing parent of finalized block."
+                            Just parent ->
+                                if isPresent (blockTimeoutCertificate block)
+                                    || isPresent (blockEpochFinalizationEntry block)
+                                    then do
+                                        -- If the block has a timeout certificate or a finalization
+                                        -- entry, then the QC on this block does not explicitly
+                                        -- finalize the parent block, so we roll back.
+                                        -- (Note, the presence of a timeout certificate indicates the
+                                        -- block is not in the next round from its parent, and the
+                                        -- presence of an epoch finalization entry indicates that block
+                                        -- is not in the same epoch as its parent. Both of these are
+                                        -- necessary for the QC to imply finalization.)
+                                        loop
+                                            (c + 1)
+                                            (finHash : hashes)
+                                            parent
+                                            (blockQuorumCertificate block)
+                                    else do
+                                        -- The block does not have a timeout certificate or finalization
+                                        -- entry, so the parent block is explicitly finalized. We thus
+                                        -- set the latest finalization entry to finalize the
+                                        -- parent block so that the database invariant is
+                                        -- restored.
+                                        storeRecord
```

Comment:
Shouldn't this be `storeReplaceRecord` rather than `storeRecord`?

I'd expect that we need to overwrite if it already exists.